compiler/ast/linter.nim

```diff
@@ -37,54 +54,119 @@ proc `=~`(s: string, a: openArray[string]): bool =
-             "range", "openarray", "varargs", "set", "cfloat", "ref", "ptr",
-             "untyped", "typed", "static", "sink", "lent", "type"]:
-      result.add s[i]
-    else:
-      result.add toUpperAscii(s[i])
-  of skConst, skEnumField:
-    # for 'const' we keep how it's spelt; either upper case or lower case:
-    result.add s[0]
+    # TODO: this should be defined elsewhere, `wordrecg`?
+    const builtIns = ["int", "int8", "int16", "int32", "int64",
+      "uint", "uint8", "uint16", "uint32", "uint64",
+      "float", "float32", "float64",
+      "cint", "cuint", "clong", "cstring",
+      "string", "char", "byte", "bool", "openArray", "seq", "array", "void",
+      "pointer", "csize", "csize_t", "cdouble", "cchar", "cschar",
+      "cshort", "cu", "nil", "typedesc", "auto", "any",
+      "range", "openarray", "varargs", "set", "cfloat", "ref", "ptr",
+      "untyped", "typed", "static", "sink", "lent", "type"]
+    for b in builtIns.items:
+      if s == b:
+        result = b
+  of skResult:
+    result = "result"
+  of skIgnore:
+    unreachable("got an impossible symbol kind: " & $k)
   else:
-    # as a special rule, don't transform 'L' to 'l'
-    if s.len == 1 and s[0] == 'L': result.add 'L'
-    elif '_' in s: result.add(s[i])
-    else: result.add toLowerAscii(s[0])
-  inc i
+    discard
+
+  if result != "":
+    return # we got a fast tracked result
+
+  var style: NamingStyle
+  case k
+  of skContainers:
+    style = lowerCamelStyle
+  of skSingletonDef:
+    style =
+      if allUpper: shoutingStyle
+      else:        upperCamelStyle
+  of skConst:
+    style = mixedCamelStyle
+  of skLiteralValue:
+    style = mixedCamelStyle
+  of skOverloadableDef:
+    style =
+      if hasUnderscore: mixedCamelStyle # `GC_` in `gc_interface`
+      elif s.len == 1:  atomStyle
+      else:             lowerCamelStyle
+  of skLocals:
+    style =
+      if s.len == 1:     atomStyle        # `L` in `lexer`
+      elif k == skParam: mixedCamelStyle  # xxx: legacy?
+      else:              lowerCamelStrict
+  of skResult, skIgnore:
+    unreachable("can never get here")
+
+  var i = 0
   while i < s.len:
-    if s[i] == '_':
-      if i+1 >= s.len:
-        discard "trailing underscores should be stripped off"
-      elif i > 0 and s[i-1] in {'A'..'Z'}:
-        # don't skip '_' as it's essential for e.g. 'GC_disable'
-        result.add('_')
-        inc i
-        result.add s[i]
+    if i == 0:    # first
+      result.add:
+        case style
+        of upperCamelStyle, shoutingStyle:    toUpperAscii(s[0])
+        of lowerCamelStyle, lowerCamelStrict: toLowerAscii(s[0])
+        of mixedCamelStyle, atomStyle:        s[0]
+    elif i+1 < s.len:
+      case s[i]
+      of '_':
+        case style
+        of upperCamelStyle, shoutingStyle, lowerCamelStyle, mixedCamelStyle:
+          result.add '_'
+          inc i
+          result.add s[i]
+        of lowerCamelStrict:
+          inc i
+          result.add toUpperAscii(s[i])
```

Comment:
It looks to be pre-existing, but doesn't the logic here mean that `a__B` gets turned into `a_B`?

Reply:
Ah, good eye. I'll leave it as is for now.

My thinking is these style rules are super capricious. Nothing really makes sense and require a ground up rethink using examples to highlight the kind of code we're want to see and the kind we don't. I got a ton of failures and the use of mixed camel case, uppercase checks, etc is me hacking around them.